lib/syntax_tools/src/erl_syntax_lib.erl

```diff
@@ -738,6 +746,44 @@ vann_binary_comp_body(Ts, Env) ->
     {Ts1, {_, Bound, Free}} = lists:mapfoldl(F, {Env, [], []}, Ts),
     {Ts1, {Bound, Free}}.
 
+vann_map_comp(Tree, Env) ->
+    Es = erl_syntax:map_comp_body(Tree),
+    {Es1, {Bound1, Free1}} = vann_map_comp_body(Es, Env),
+    Env1 = ordsets:union(Env, Bound1),
+    T = erl_syntax:map_comp_template(Tree),
+    {T1, _, Free2} = vann(T, Env1),
+    Free = ordsets:union(Free1, ordsets:subtract(Free2, Bound1)),
+    Bound = [],
+    Tree1 = rewrite(Tree, erl_syntax:map_comp(T1, Es1)),
+    {ann_bindings(Tree1, Env, Bound, Free), Bound, Free}.
+
+vann_map_comp_body_join() ->
+    fun (T, {Env, Bound, Free}) ->
+            {T1, Bound1, Free1} = case erl_syntax:type(T) of
+                                      binary_generator ->
+				          vann_binary_generator(T,Env);
+				      generator ->
+                                          vann_generator(T, Env);
+                                      map_generator ->
+                                          vann_map_generator(T,Env);
+                                      _ ->
+                                          %% Bindings in filters are not
+                                          %% exported to the rest of the
+                                          %% body.
+                                          {T2, _, Free2} = vann(T, Env),
+                                          {T2, [], Free2}
+                                  end,
+            Env1 = ordsets:union(Env, Bound1),
+            {T1, {Env1, ordsets:union(Bound, Bound1),
+                  ordsets:union(Free, 
+                                ordsets:subtract(Free1, Bound))}}
+    end.
```

Comment:
Stylish question, there are now three (3) copies of this function (`vann_map_comp_body_join`, `vann_list_comp_body_join` and `vann_binary_comp_body_join` where the last one has a slightly different indentation!) they include nothing map/list/binary specific, could they be replaced by a single `vann_comp_body_join`? That should make future updates less error prone? 🤔